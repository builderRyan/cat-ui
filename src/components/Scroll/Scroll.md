# Scroll 滚动

## Description

该组件提供子内容滚动位置距离窗口顶部的的实时监测，并提供了回调。

## Example

```jsx
<Scroll onPageIndexChange={onPageIndexChange}>
  <div className='block-1'></div>
  <div className='block-2'></div>
  <div className='block-3'></div>
  <div className='block-4'></div>
</Scroll>
```

## API

|       属性        |        类型         |       默认值       |       描述        | 是否必填 |
| :---------------: | :-----------------: | :----------------: | :---------------: | :------: |
|    targetNode     |      `string`       | `target-dom-point` |     目标节点      |    -     |
|      offset       |      `number`       |        `0`         |      偏移量       |    -     |
|      prefix       |      `string`       |    `cat-scroll`    |     类名前缀      |    -     |
|     className     |      `string`       |       `cat`        |       类名        |    -     |
| onPageIndexChange |       `func`        |         -          | page 切换时的回调 |    -     |
|       style       |      `object`       |         -          |       样式        |    -     |
|     children      | `React.ReactNode[]` |         -          |     children      |    -     |
|       wait        |      `number`       |       `250`        |  节流器等待时间   |    -     |
