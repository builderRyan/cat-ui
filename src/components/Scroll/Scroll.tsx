import * as React from 'react';
import * as PropTypes from 'prop-types';
import throttle from "lodash/throttle";
import classNames from 'classnames';

interface ScrollProps {
  targetNode?: string;
  offset?: number;
  prefix?: string;
  className?: string;
  onPageIndexChange?: (index: number) => void;
  style?: object;
  children: React.ReactNode[];
  wait?: number;
}

type PageHeight = {
  height: number,
  index: number
}

export default class Scroll extends React.Component<ScrollProps, {}> {
  static propTypes = {
    targetNode: PropTypes.string,
    offset: PropTypes.number,
    prefix: PropTypes.string,
    className: PropTypes.string,
    onPageIndexChange: PropTypes.func,
    style: PropTypes.object,
    children: PropTypes.array.isRequired,
    wait: PropTypes.number,
  };


  static defaultProps = {
    prefix: 'cat-scroll',
    targetNode: "target-dom-point",
    offset: 0,
    wait: 250
  };

  private targetNode: string;
  private pageHeights: PageHeight[];

  constructor(props) {
    super(props);
    this.targetNode = this.props.targetNode;
    this.scrollHandler = this.scrollHandler.bind(this);
    this.initPageHeight = this.initPageHeight.bind(this);
  }

  componentDidMount() {
    this.initPageHeight();
    window.addEventListener("scroll", this.scrollHandler, {
      passive: true
    });
  }

  componentWillUnmount() {
    window.removeEventListener("scroll", this.scrollHandler);
  }

  initPageHeight() {
    const targetNodes = document.querySelectorAll(`.${this.targetNode}`);
    const targetHeights = Array.from(targetNodes).map(node => {
      return node.getBoundingClientRect().height;
    });

    let sum = 0;
    this.pageHeights = targetHeights.map((height, index) => {
      sum += height;
      return {
        height: sum,
        index: index + 1
      };
    });
  }

  scrollHandler = throttle(() => {
    const { offset, onPageIndexChange } = this.props;

    const scrollTop =
      document.documentElement.scrollTop || document.body.scrollTop;
    let pageIndex = 0;
    this.pageHeights.forEach(page => {
      if (scrollTop + offset > page.height) {
        pageIndex = page.index;
      }
    });
    onPageIndexChange && onPageIndexChange(pageIndex);
  }, this.props.wait);

  render() {
    const {
      prefix,
      className,
      style,
    } = this.props;

    const classes = classNames(prefix, className);

    const children = this.props.children.map((child, index) => {
      return (
        <li className={this.targetNode} key={index}>
          {child}
        </li>
      );
    });
    return <ul className={classes} prefix={prefix} style={style}>{children}</ul>;
  }
}
