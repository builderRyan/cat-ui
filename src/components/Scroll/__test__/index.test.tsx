import * as React from 'react'
import * as Enzyme from 'enzyme'
import { render, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import Scroll from '../index'

Enzyme.configure({ adapter: new Adapter() })

describe('Scroll', () => {
  test('renders correctly', () => {
    const wrapper = render(
      <Scroll>
        <div>block1</div>
        <div>block2</div></Scroll>
    )
    expect(wrapper).toMatchSnapshot()
  })

})
