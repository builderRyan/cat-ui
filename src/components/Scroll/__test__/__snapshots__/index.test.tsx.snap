// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scroll renders correctly 1`] = `
<ul
  class="cat-scroll"
  prefix="cat-scroll"
>
  <li
    class="target-dom-point"
  >
    <div>
      block1
    </div>
  </li>
  <li
    class="target-dom-point"
  >
    <div>
      block2
    </div>
  </li>
</ul>
`;
